### PR TITLE
Rewrite downloading the nuget dependencies

### PR DESCRIPTION
**Pull request by annakocheshkova**
Things to consider before you submit the PR:

* [X] Are the files formatted correctly?
* [X] Did you test your change?

## Description

Rewrite downloading the nuget dependencies.

## Related PRs or issues

[AB#77174](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/77174)

### CODE REVIEW

File: Assets/AppCenter/Editor/AppCenterPreBuild.cs [hunk 1]
```diff
@@ -47,6 +47,28 @@ public void OnPreprocessBuild(BuildTarget target, string path)
 #if !APPCENTER_DONT_USE_NATIVE_STARTER
             AddStartupCode(new AppCenterSettingsMakerIos());
 #endif
+        } else if (target == BuildTarget.WSAPlayer) {
+            CleanOldDependencies();
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPreBuild.cs]: I guess it will be resolve by default .unitypackage importing logic

File: build.cake [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
+using System.Threading;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Resolver;
+using NuGet.Versioning;
```
Comment (MatkovIvan) [inline build.cake]: Please move all manual NuGet operations to separate cake file

File: build.cake [hunk 3]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
 }).OnError(HandleError);
 
-// Downloading UWP IL2CPP dependencies.
+
+async Task<SourcePackageDependencyInfo> ResolvePackage(DependencyInfoResource dependencyResource, PackageIdentity identity) 
+{
+    var cacheContext = new SourceCacheContext();
+    var logger = NullLogger.Instance;
+
+    var package = await dependencyResource.ResolvePackage(identity, Netstandard2FrameworkName, cacheContext, logger, CancellationToken.None);
```
Comment (MatkovIvan) [inline build.cake]: UWP must be checked first

File: build.cake [hunk 4]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
+{
+    Information($"GetRecursiveDependenciesCore {id} version {version.Version.ToString()}");
+    var package = await ResolvePackage(dependencyResource, new PackageIdentity(id, version));
+    dependencies.Add(new NugetDependency(package.Id, NuGetVersion.Parse(package.Version.ToNormalizedString())));
+    var hashPackageId = GetStringHash(package.Id);
+    if (!uniqueIds.Contains(hashPackageId))
+    {
+        uniqueIds.Add(hashPackageId);
+        var uri = package.DownloadUri.ToString();
+        Information("Downloading " + package.Id + " from " + uri);
```
Comment (MatkovIvan) [inline build.cake]: Downloading in "Get*" function? please separate the logic

File: build.cake [hunk 5]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
+        DownloadFile (uri, ExternalsFolder + package.Id + NuPkgExtension);
+        if (!addDependencies) 
+        {
+            return;
+        }
+        foreach (var dependency in package.Dependencies) 
+        {
+            if (IgnoreNuGetDependencies.Contains(dependency.Id) || dependency.Id.StartsWith("System"))
```
Comment (MatkovIvan) [inline build.cake]: Move it to `IgnoreNuGetDependencies` list, or use patterns

File: build.cake [hunk 6]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
+        {
+            return;
+        }
+        foreach (var dependency in package.Dependencies) 
+        {
+            if (IgnoreNuGetDependencies.Contains(dependency.Id) || dependency.Id.StartsWith("System")) 
+            {
+                continue;
+            }
+            var hashDepId = GetStringHash(dependency.Id);
```
Comment (MatkovIvan) [inline build.cake]: Why don't use strings directly?

File: build.cake [hunk 7]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
+    } 
+}
+
+private int GetStringHash(string input) 
+{
+    using (System.Security.Cryptography.HashAlgorithm algorithm = System.Security.Cryptography.SHA256.Create()) 
+    {
+        return BitConverter.ToInt32(algorithm.ComputeHash(Encoding.UTF8.GetBytes(input)), 0);
+    }
+}
```
Comment (MatkovIvan) [inline build.cake]: Remove it

File: build.cake [hunk 8]
```diff
@@ -378,36 +401,98 @@ Task("Install-Unity-Windows").Does(() => {
...
+
+private DependencyInfoResource GetDefaultDependencyResource() 
+{
+    var packageSource = "https://api.nuget.org/v3/index.json";
```
Comment (MatkovIvan) [inline build.cake]: Use built in const

File: build.cake [hunk 9]
```diff
@@ -554,38 +639,45 @@ Task("DownloadNdk")
...
 
-void GetUwpPackage (AppCenterModule module, bool usePublicFeed) {
+async Task GetUwpPackage (AppCenterModule module, bool usePublicFeed) {
     // Prepare destination
     var destination = "Assets/AppCenter/Plugins/WSA/" + module.Moniker + "/";
     EnsureDirectoryExists (destination);
     DeleteFiles (destination + "*.dll");
     DeleteFiles (destination + "*.winmd");
     var nupkgPath = "externals/uwp/";
+    Console.WriteLine($"GetUwpPackage use public feed {usePublicFeed}");
```
Comment (MatkovIvan) [inline build.cake]: Use cake logger

File: build.cake [hunk 10]
```diff
@@ -554,38 +639,45 @@ Task("DownloadNdk")
...
     if (usePublicFeed) {
-        var packageSource = "https://www.nuget.org/api/v2/";
-        var repository = PackageRepositoryFactory.Default.CreateRepository (packageSource);
-
-        var package = repository.FindPackage (module.DotNetModule, SemanticVersion.Parse (UwpSdkVersion));
-        IEnumerable<IPackage> dependencies = new [] { package };
-        var fileSystem = new PhysicalFileSystem (Environment.CurrentDirectory);
+        List<NugetDependency> dependencies = new List<NugetDependency>();
+        HashSet<int> uniqueIds = new HashSet<int>();
+        var dep = new NugetDependency(module.DotNetModule, UwpSdkVersion);
```
Comment (MatkovIvan) [inline build.cake]: Please use full names

File: build.cake [hunk 11]
```diff
@@ -554,38 +639,45 @@ Task("DownloadNdk")
...
     var nupkgPath = "externals/uwp/";
+    Console.WriteLine($"GetUwpPackage use public feed {usePublicFeed}");
     if (usePublicFeed) {
-        var packageSource = "https://www.nuget.org/api/v2/";
-        var repository = PackageRepositoryFactory.Default.CreateRepository (packageSource);
-
-        var package = repository.FindPackage (module.DotNetModule, SemanticVersion.Parse (UwpSdkVersion));
-        IEnumerable<IPackage> dependencies = new [] { package };
-        var fileSystem = new PhysicalFileSystem (Environment.CurrentDirectory);
+        List<NugetDependency> dependencies = new List<NugetDependency>();
```
Comment (MatkovIvan) [inline build.cake]: `var`s

File: build.cake [hunk 12]
```diff
@@ -796,72 +888,91 @@ string GetNuGetPackage(string packageId, string packageVersion)
...
+    {
+        return files;
+    }
+    var netStandard2Path = $"{tempContentPath}/lib/{Netstandard2FrameworkName.GetShortFolderName()}/*.dll";
+    files = GetFiles(netStandard2Path);
+    if (files.Any()) 
+    {
+        return files;
+    }
+    if (files.Any())
```
Comment (MatkovIvan) [inline build.cake]: ?

File: build.cake [hunk 13]
```diff
@@ -796,72 +888,91 @@ string GetNuGetPackage(string packageId, string packageVersion)
...
-    var contentPathSuffix = "lib/uap10.0/";
+    Warning($"Haven't found anything under {tempContentPath} - make sure it's expected.");
+    return null;
+}
+
+string MoveNativeBinaries(string tempContentPath, string destination)
+{
     var runtimesPath = tempContentPath + "/runtimes";
     if (DirectoryExists (runtimesPath)) {
+        Console.WriteLine("MoveNativeBinaries");
```
Comment (MatkovIvan) [inline build.cake]: Cake logger

File: build.cake [hunk 14]
```diff
@@ -796,72 +888,91 @@ string GetNuGetPackage(string packageId, string packageVersion)
...
         var oneArch = "x86";
         foreach (var runtime in GetDirectories (runtimesPath + "/win10-*")) {
             var arch = runtime.GetDirectoryName ().ToString ().Replace ("win10-", "").ToUpper ();
             oneArch = arch;
-            var nativeFiles = GetFiles (runtime + "/native/*");
+            var nativeFiles = GetFiles (runtime + "/**/*.dll");
```
Comment (MatkovIvan) [inline build.cake]: Why did you change it?

File: build.cake [hunk 15]
```diff
@@ -796,72 +888,91 @@ string GetNuGetPackage(string packageId, string packageVersion)
...
             var targetArchPath = destination + "/" + arch;
             EnsureDirectoryExists (targetArchPath);
             foreach (var nativeFile in nativeFiles) {
                 if (!FileExists (targetArchPath + "/" + nativeFile.GetFilename ())) {
                     MoveFileToDirectory (nativeFile, targetArchPath);
+                    Console.WriteLine($"Moved native binary file {nativeFile} to {targetArchPath}");
+                } else {
+                    Console.WriteLine("Native binary file already exists");
```
Comment (MatkovIvan) [inline build.cake]: Cake Logger

File: build.cake [hunk 16]
```diff
@@ -796,72 +888,91 @@ string GetNuGetPackage(string packageId, string packageVersion)
...
+    var packageName = package != null ? package.Name : "null";
+    Console.WriteLine($"ExtractNuGetPackage {packageName}; tempcontentpath {tempContentPath}; destination {destination}");
+    var oneArch = MoveNativeBinaries(tempContentPath, destination);
+    if (package != null) {
+        MoveAssemblies(package, tempContentPath, destination);
+    }
+    if (package == null) {
+        var contentPathSuffix = "lib/uap10.0/";
+        if (oneArch != null) {
+            //todo check, the result path may be incorrect
```
Comment (MatkovIvan) [inline build.cake]: TODO

File: packageextractor.cake [hunk 17]
```diff
@@ -0,0 +1,99 @@
+class PackageExtractor
```
Comment (MatkovIvan) [inline packageextractor.cake]: Copyright

File: packageextractor.cake [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+{
+    static PackageSaveMode packageSaveMode = PackageSaveMode.Defaultv3;
+
+    public static void Extract(string fileName)
+    {
+        string dir = Environment.CurrentDirectory;
```
Comment (MatkovIvan) [inline packageextractor.cake]: var

File: packageextractor.cake [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        using (var stream = new FileStream(fileName, FileMode.Open, FileAccess.Read))
+        {
+            Extract(stream, targetPath);
+        }
+    }
+
+    public static void Extract(Stream nupkgStream, string targetPath)
+    {
+        var tempHashPath = System.IO.Path.Combine(targetPath, System.IO.Path.GetRandomFileName());
```
Comment (MatkovIvan) [inline packageextractor.cake]: `using System.IO;`

File: packageextractor.cake [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+        var targetNuspec = System.IO.Path.Combine(targetPath, System.IO.Path.GetRandomFileName());
+        var targetNupkg = System.IO.Path.Combine(targetPath, System.IO.Path.GetRandomFileName());
+        targetNupkg = System.IO.Path.ChangeExtension(targetNupkg, ".nupkg");
```
Comment (MatkovIvan) [inline packageextractor.cake]: I saw const for this ext

File: packageextractor.cake [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        var hashPath = System.IO.Path.Combine(targetPath, System.IO.Path.GetRandomFileName());
+
+        using (var packageReader = new PackageArchiveReader(nupkgStream))
+        {
+            var nuspecFile = packageReader.GetNuspecFile();
+            if ((packageSaveMode & PackageSaveMode.Nuspec) == PackageSaveMode.Nuspec)
+            {
+                packageReader.ExtractFile(nuspecFile, targetNuspec, new NullLogger());
```
Comment (MatkovIvan) [inline packageextractor.cake]: Why NullLogger? there was some global context or so

File: packageextractor.cake [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+            string packageHash;
+            nupkgStream.Position = 0;
+            packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(nupkgStream));
+
+            System.IO.File.WriteAllText(tempHashPath, packageHash);
+        }
+    }
+
+    private static bool ShouldInclude(
```
Comment (MatkovIvan) [inline packageextractor.cake]: Why not just *.dll?

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: build.cake (hunk 12 above)
Comment (annakocheshkova) [reply build.cake]: This is to check if enumerable is not empty.

File: build.cake (hunk 14 above)
Comment (annakocheshkova) [reply build.cake]: @russelarms

Comment (russelarms) [reply build.cake]: The script looked for files in locations like `externals/uwp/SQLitePCLRaw.lib.e_sqlite3/runtimes/win10-arm64/native/*`
But now files for win10 are located in
`externals/uwp/SQLitePCLRaw.lib.e_sqlite3/runtimes/win10-arm64/nativeassets/uap10.0/e_sqlite3.dll`
So it doesn't work without that change.

File: build.cake (hunk 16 above)
Comment (russelarms) [reply build.cake]: I checked but didn't find a case where `oneArch` is set to something else with our current set of libraries.
A different set of runtime DLLs are used only in our SQLite lib for now, which is used from a public source only.
cc @annakocheshkova

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs [hunk 23]
```diff
@@ -312,10 +312,9 @@ private static void AddDependenciesToProjectJson(string projectJsonPath)
             return;
         }
         var jsonString = File.ReadAllText(projectJsonPath);
-        jsonString = AddDependencyToProjectJson(jsonString, "Microsoft.NETCore.UniversalWindowsPlatform", "5.2.2");
-        jsonString = AddDependencyToProjectJson(jsonString, "Newtonsoft.Json", "10.0.3");
-        jsonString = AddDependencyToProjectJson(jsonString, "sqlite-net-pcl", "1.3.1");
-        jsonString = AddDependencyToProjectJson(jsonString, "System.Collections.NonGeneric", "4.0.1");
+        jsonString = AddDependencyToProjectJson(jsonString, "Microsoft.NETCore.UniversalWindowsPlatform", "6.2.9");
+        jsonString = AddDependencyToProjectJson(jsonString, "Newtonsoft.Json", "12.0.3");
+        jsonString = AddDependencyToProjectJson(jsonString, "SQLitePCLRaw.bundle_green", "2.0.2");
```
Comment (MatkovIvan) [inline Assets/AppCenter/Editor/AppCenterPostBuild.cs]: > You can't use .NET Standard libraries with .NET scripting backend. You'll have to switch to IL2CPP.

https://forum.unity.com/threads/uwp-app-with-netstandard-libraries.531751/#post-3499618

Does this really work now?

File: build.cake (hunk 8 above)
Comment (russelarms) [reply build.cake]: I wasn't able to make use of [CAKE_NUGET_SOURCE](https://cakebuild.net/docs/fundamentals/default-configuration-values), so I just moved it to the top as a constant.

File: packageextractor.cake (hunk 21 above)
Comment (russelarms) [reply packageextractor.cake]: This class is from another inheritance tree 
`NuGet.Common.ILogger`.

File: packageextractor.cake (hunk 20 above)
Comment (russelarms) [reply packageextractor.cake]: The const is from the main file.

File: build.cake [hunk 24]
```diff
@@ -378,37 +369,18 @@ Task("Install-Unity-Windows").Does(() => {
...
 
-// Downloading UWP IL2CPP dependencies.
+//Downloading UWP IL2CPP dependencies.
 Task ("Externals-Uwp-IL2CPP-Dependencies")
-    .Does (() => {
+    .Does (async () => {
         var targetPath = "Assets/AppCenter/Plugins/WSA/IL2CPP";
         EnsureDirectoryExists (targetPath);
+        EnsureDirectoryExists (ExternalsFolder);
         EnsureDirectoryExists (targetPath + "/ARM");
```
Comment (MatkovIvan) [inline build.cake:369]: Add ARM64

File: build.cake (hunk 2 above)
Comment (MatkovIvan) [reply build.cake]: And remove unused usings

File: build.cake [hunk 25]
```diff
@@ -554,41 +526,26 @@ Task("DownloadNdk")
...
 
-void GetUwpPackage (AppCenterModule module, bool usePublicFeed) {
+async Task GetUwpPackage (AppCenterModule module, bool usePublicFeed) {
     // Prepare destination
     var destination = "Assets/AppCenter/Plugins/WSA/" + module.Moniker + "/";
     EnsureDirectoryExists (destination);
     DeleteFiles (destination + "*.dll");
     DeleteFiles (destination + "*.winmd");
     var nupkgPath = "externals/uwp/";
+    Information($"GetUwpPackage use public feed {usePublicFeed}");
```
Comment (MatkovIvan) [inline build.cake]: Move this inside if to use more accurate wording. Also, please avoid usage internal method names in logs

File: nuget-tools.cake [hunk 26]
```diff
@@ -0,0 +1,173 @@
+var NugetUrl = "https://api.nuget.org/v3/index.json";
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Should be removed. Use built-in const.

Comment (MatkovIvan) [inline nuget-tools.cake]: Move nuget library includes here

File: nuget-tools.cake [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+// UWP IL2CPP dependencies.
+var UwpIL2CPPDependencies = new [] {
+   new NugetDependency("SQLitePCLRaw.bundle_green", "2.0.2", ".NETStandard,Version=v2.0"),
+   new NugetDependency("Newtonsoft.Json", "12.0.3", ".NETStandard,Version=v2.0"),
+   new NugetDependency("SQLitePCLRaw.provider.e_sqlite3", "2.0.2", ".NETStandard,Version=v2.0"),
+   new NugetDependency("SQLitePCLRaw.core", "2.0.2", ".NETStandard,Version=v2.0"),
+   new NugetDependency("SQLitePCLRaw.lib.e_sqlite3", "2.0.2", "UAP10.0")
+};
+
+var UwpIL2CPPJsonUrl = SdkStorageUrl + "Newtonsoft.Json.dll";
```
Comment (MatkovIvan) [inline nuget-tools.cake:27]: Why do you need both NugetDependency and UwpIL2CPPJsonUrl for it?

File: nuget-tools.cake [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+    return sourceRepository.GetResource<DependencyInfoResource>(CancellationToken.None);
+}
+
+async Task ProcessDependency(NugetDependency dependency, string destination) 
+{
+    var package = await GetDependency(GetDefaultDependencyResource(), dependency);
+    var uri = package.DownloadUri.ToString();
+    Information("Downloading " + package.Id + " from " + uri);
+    DownloadFile (uri, ExternalsFolder + package.Id + NuPkgExtension);
+    Information ("Extract NuGet package: " + dependency.Name);
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Format

File: nuget-tools.cake [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+            }
+        }
+        return oneArch;
+    } 
+    return null;
+}
+
+void MoveAssemblies(NugetDependency package, string tempContentPath, string destination)
+{
+    Information($"MoveAssemblies");
```
Comment (MatkovIvan) [inline nuget-tools.cake]: for what package? to where?

File: nuget-tools.cake [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+    if (package != null) 
+    {
+        MoveAssemblies(package, tempContentPath, destination);
+    }
+    if (package == null) 
+    {
+        var contentPathSuffix = "lib/uap10.0/";
+        if (oneArch != null) 
+        {
+            //todo check, the result path may be incorrect
```
Comment (MatkovIvan) [inline nuget-tools.cake]: TODO

File: nuget-tools.cake (hunk 26 above)
Comment (MatkovIvan) [inline nuget-tools.cake]: Copyright?

File: packageextractor.cake [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+                    packageFileExtractor.ExtractPackageFile,
+                    new NullLogger(),
+                    CancellationToken.None);
+            }
+
+            string packageHash;
+            nupkgStream.Position = 0;
+            packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(nupkgStream));
+
+            System.IO.File.WriteAllText(tempHashPath, packageHash);
```
Comment (MatkovIvan) [inline packageextractor.cake]: Using

File: packageextractor.cake [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+                    targetPath,
+                    packageFiles,
+                    packageFileExtractor.ExtractPackageFile,
+                    new NullLogger(),
+                    CancellationToken.None);
+            }
+
+            string packageHash;
+            nupkgStream.Position = 0;
+            packageHash = Convert.ToBase64String(new CryptoHashProvider("SHA512").CalculateHash(nupkgStream));
```
Comment (MatkovIvan) [inline packageextractor.cake]: Why do you need it?

File: nuget-tools.cake [hunk 33]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+string GetNuGetPackage(string packageId, string packageVersion)
+{
+    var nugetUser = EnvironmentVariable("NUGET_USER");
+    var nugetPassword = Argument("NuGetPassword", EnvironmentVariable("NUGET_PASSWORD"));
+    var nugetFeedId = Argument("NuGetFeedId", EnvironmentVariable("NUGET_FEED_ID"));
+    packageId = packageId.ToLower();
+
+    var url = "https://msmobilecenter.pkgs.visualstudio.com/_packaging/";
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Move to const

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs (hunk 23 above)
Comment (MatkovIvan) [reply Assets/AppCenter/Editor/AppCenterPostBuild.cs]: Also, this should be in a separate PR

File: nuget-tools.cake (hunk 27 above)
Comment (annakocheshkova) [reply nuget-tools.cake:27]: Because removing this is addressed in another PR.

File: Assets/AppCenter/Editor/AppCenterPostBuild.cs (hunk 23 above)
Comment (annakocheshkova) [reply Assets/AppCenter/Editor/AppCenterPostBuild.cs]: > Does this really work now?

This is related to E2E task. Also, we can't fully check this due to another blocking issue which is solved in another task.

File: build.cake [hunk 34]
```diff
@@ -1,20 +1,27 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
 #addin nuget:?package=Cake.FileHelpers
 #addin nuget:?package=Cake.AzureStorage
-#addin nuget:?package=NuGet.Core
+#addin nuget:?package=NuGet.Resolver&loaddependencies=true
+#addin nuget:?package=NuGet.Protocol&loaddependencies=true
```
Comment (MatkovIvan) [inline build.cake]: Move to Nuget related cake file

File: build.cake [hunk 35]
```diff
@@ -378,43 +363,21 @@ Task("Install-Unity-Windows").Does(() => {
...
-            foreach (var depPackage in dependencies) {
-                Information ("Extract NuGet package: " + depPackage);
-
-                // Extract.
-                var path = "externals/uwp/" + depPackage.Id;
-                depPackage.ExtractContents (fileSystem, path);
-
-                ExtractNuGetPackage (depPackage, frameworkName, path, targetPath);
-            }
+        EnsureDirectoryExists (targetPath + "/ARM64");
```
Comment (MatkovIvan) [inline build.cake]: Move this line to another ARM line

File: nuget-tools.cake [hunk 36]
```diff
@@ -0,0 +1,170 @@
...
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+var PrivateNugetFeedUrl = "https://msmobilecenter.pkgs.visualstudio.com/_packaging/";
+
+var UwpIL2CPPDependencies = new [] {
```
Comment (MatkovIvan) [inline nuget-tools.cake:20]: Add comment

File: nuget-tools.cake [hunk 37]
```diff
@@ -0,0 +1,170 @@
...
+};
+
+var UwpIL2CPPJsonUrl = SdkStorageUrl + "Newtonsoft.Json.dll";
+var ExternalsFolder = "externals/uwp/";
+var NuPkgExtension = ".nupkg";
+
+var IgnoreNuGetDependencies = new[] {
+    "Microsoft.NETCore.UniversalWindowsPlatform",
+    "NETStandard.Library"
+};
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Is it used?

File: nuget-tools.cake [hunk 38]
```diff
@@ -0,0 +1,170 @@
...
+    return sourceRepository.GetResource<DependencyInfoResource>(CancellationToken.None);
+}
+
+async Task ProcessDependency(NugetDependency dependency, string destination) 
+{
+    var package = await GetDependency(GetDefaultDependencyResource(), dependency);
+    var uri = package.DownloadUri.ToString();
+    Information("Downloading " + package.Id + " from " + uri);
+    DownloadFile (uri, ExternalsFolder + package.Id + NuPkgExtension);
+    Information ($"Extract NuGet package: {dependency.Name}");
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Format

File: nuget-tools.cake [hunk 39]
```diff
@@ -0,0 +1,170 @@
...
+    packageId = packageId.ToLower();
+    var url = PrivateNugetFeedUrl + nugetFeedId + "/nuget/v3/flat2/" + packageId + "/" + packageVersion + "/" + packageId + "." + packageVersion + ".nupkg";
+
+    // Get the NuGet package
+    HttpWebRequest request = (HttpWebRequest)WebRequest.Create (url);
+    request.Headers["X-NuGet-ApiKey"] = nugetPassword;
+    request.Credentials = new NetworkCredential(nugetUser, nugetPassword);
+    HttpWebResponse response = (HttpWebResponse)request.GetResponse();
+    var responseString = String.Empty;
+    var filename = packageId + "." + packageVersion +  ".nupkg";
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Reuse in url

File: nuget-tools.cake (hunk 29 above)
Comment (MatkovIvan) [reply nuget-tools.cake]: It not resolved

File: nuget-tools.cake [hunk 40]
```diff
@@ -0,0 +1,170 @@
...
+        if (FileExists(targetPath)) 
+        {
+            DeleteFile(targetPath);
+        }
+        MoveFile(matchingFile.FullPath, targetPath);
+        Information($"Moving {matchingFile.FullPath} to {targetPath}");
+    } 
+}
+
+void ExtractNuGetPackage (NugetDependency package, string tempContentPath, string destination)
```
Comment (MatkovIvan) [inline nuget-tools.cake]: Why is it not in extractor? Why do you need 2 files for nuget tools?

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs [hunk 41]
```diff
@@ -209,13 +209,19 @@ private static void Prepare()
...
                     UnityApplicationSettings.Initialize();
                     UWPAppCenter.SetApplicationSettingsFactory(new UnityApplicationSettingsFactory());
 
                     /**
                      * Workaround for another IL2CPP issue.
                      * System.Net.Http.HttpClient doesn't work properly so, replace to unity specific implementation.
+                     * I found various related issues in tracker, with different Unity versions marked as ones which fixed the issue:
+                     * https://issuetracker.unity3d.com/issues/reference-to-httpclient-statement-is-not-found-in-vs2017
+                     * https://issuetracker.unity3d.com/issues/httpclient-namespace-is-not-recognized-in-vs-with-net-4-dot-x
+                     * Can be removed after minimum unity version is bumped, and after verifying that it is actually fixed.
```
Comment (MatkovIvan) [inline Assets/AppCenter/Plugins/AppCenterSDK/Core/UWP/AppCenterInternal.cs]: This should not be a part of this PR

File: nuget-tools.cake [hunk 42]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+async Task ProcessDependency(NugetDependency dependency, string destination) 
+{
+    var package = await GetDependency(GetDefaultDependencyResource(), dependency);
+    var uri = package.DownloadUri.ToString();
+    Information($"Downloading {package.Id} from {uri}");
+    DownloadFile(uri, ExternalsFolder + package.Id + ".nupkg");
+    Information($"Extract NuGet package: {dependency.Name}");
+    var path = ExternalsFolder + dependency.Name + ".nupkg";
```
Comment (MatkovIvan) [inline nuget-tools.cake:73]: Can it be used in 71 line?

File: nuget-tools.cake [hunk 43]
```diff
@@ -0,0 +1,219 @@
...
+{
+    var nugetUser = EnvironmentVariable("NUGET_USER");
+    var nugetPassword = Argument("NuGetPassword", EnvironmentVariable("NUGET_PASSWORD"));
+    var nugetFeedId = Argument("NuGetFeedId", EnvironmentVariable("NUGET_FEED_ID"));
+    packageId = packageId.ToLower();
+    var filename = packageId + "." + packageVersion +  ".nupkg";
+    var url = $"{PrivateNugetFeedUrl}{nugetFeedId}/nuget/v3/flat2/{packageId}/{packageVersion}/{filename}";
+
+    // Get the NuGet package
+    HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
```
Comment (MatkovIvan) [inline nuget-tools.cake:89]: var?

File: nuget-tools.cake [hunk 44]
```diff
@@ -0,0 +1,219 @@
...
+        return files;
+    }
+    Warning($"Haven't found anything under {tempContentPath} - make sure it's expected.");
+    return null;
+}
+
+void MoveNativeBinaries(string tempContentPath, string destination)
+{
+    var runtimesPath = tempContentPath + "/runtimes";
+    if (DirectoryExists(runtimesPath))
```
Comment (MatkovIvan) [inline nuget-tools.cake:116]: Let's revert the condition and return. This will reduce indent and improve readability

File: nuget-tools.cake [hunk 45]
```diff
@@ -0,0 +1,219 @@
...
+            DeleteFile(targetPath);
+        }
+        MoveFile(matchingFile.FullPath, targetPath);
+        Information($"Moving {matchingFile.FullPath} to {targetPath}");
+    } 
+}
+
+void ProcessPackageDlls(NugetDependency package, string tempContentPath, string destination)
+{
+    Information($"ProcessPackageDlls for {package.Name}; tempcontentpath {tempContentPath}; destination {destination}");
```
Comment (MatkovIvan) [inline nuget-tools.cake:160]: Avoid internal method names in logs. Just writes what's happening here

File: nuget-tools.cake [hunk 46]
```diff
@@ -0,0 +1,219 @@
...
+    MoveNativeBinaries(tempContentPath, destination);
+    MoveAssemblies(package, tempContentPath, destination);
+}
+
+void ProcessAppCenterDlls(string tempContentPath, string destination) 
+{
+    MoveNativeBinaries(tempContentPath, destination);
+    var contentPathSuffix = "lib/uap10.0/";
+    Information($"Moving SDK package, move from {tempContentPath + contentPathSuffix + '*'} to {destination}");
+    var files = GetFiles(tempContentPath + contentPathSuffix + "*");
```
Comment (MatkovIvan) [inline nuget-tools.cake:170]: Can be used in the log above

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: nuget-tools.cake (hunk 43 above)
Comment (MatkovIvan) [reply nuget-tools.cake:89]: https://cakebuild.net/api/Cake.Common.Net/HttpAliases/9E96C418 ?

Comment (annakocheshkova) [reply nuget-tools.cake:89]: Why? There's credentials + header value below.

**Comment by annakocheshkova** [comment]
PR is cleaned up in #418
